### PR TITLE
Handle panics in test runtime

### PR DESCRIPTION
**Pull request by alfonsoros88**

### CODE REVIEW

File: northstar/src/main.rs [hunk 1]
```diff
@@ -80,26 +80,26 @@ fn main() -> Result<(), Error> {
...
         _ = sigterm.recv() => {
             info!("Received SIGTERM. Stopping Northstar runtime");
-            runtime.stop_wait().await
+            runtime.stop()
         }
-        status = &mut runtime => status,
     };
-    match status {
+
+    match runtime.stop_wait().await {
```
Comment (flxo) [inline northstar/src/main.rs:102]: Why is this change needed? A `stop_wait` does the same as a `stop` and `stop_wait.await`.

File: northstar_tests/src/runtime.rs [hunk 2]
```diff
@@ -65,7 +65,16 @@ pub fn timeout_on<R>(f: impl Future<Output = R>) -> Result<R> {
 }
 
 /// A running instance of northstar.
-pub struct Runtime(runtime::Runtime);
+pub struct Runtime(Option<runtime::Runtime>);
+
+impl Drop for Runtime {
```
Comment (flxo) [inline northstar_tests/src/runtime.rs:70]: drop takes a &mut self. Why is this needed to be an `Option`? Because you want to move it into the `block_on`?

File: northstar/src/runtime/mod.rs [hunk 3]
```diff
@@ -94,7 +89,7 @@ pub struct Runtime {
 
 impl Runtime {
     pub async fn start(config: Config) -> Result<Runtime, Error> {
-        let (stop_tx, stop_rx) = oneshot::channel();
+        let token = CancellationToken::new();
```
Comment (flxo) [inline northstar/src/runtime/mod.rs:92]: Nice.

File: northstar_tests/src/runtime.rs (hunk 2 above)
Comment (alfonsoros88) [reply northstar_tests/src/runtime.rs:70]: Yes, exactly this :arrow_up:

File: northstar/src/main.rs (hunk 1 above)
Comment (alfonsoros88) [reply northstar/src/main.rs:102]: Hmmm yes, I think this is wrong :thinking:

**Review by flxo — APPROVED** [review]

**Comment by flxo** [comment]
@alfonsoros88 Let's talk about `Runtime::stop` and `Runtime::wait`. I'd merge them again. because stopping without waiting for a result doesn't make sense and ignore the result is always an option.

File: northstar/src/runtime/mod.rs [hunk 4]
```diff
@@ -118,23 +121,20 @@ impl Runtime {
...
+            token,
+        ))
     }
 
-    /// Stop the runtime and wait for the termination
-    pub fn stop_wait(mut self) -> impl Future<Output = RuntimeResult> {
-        self.stop.take();
-        self
+    /// Wait for the termination
+    pub async fn wait(self) -> RuntimeResult {
```
Comment (flxo) [inline northstar/src/runtime/mod.rs:134]: Should me merged into `Runtime::stop`

File: northstar/src/runtime/mod.rs [hunk 5]
```diff
@@ -93,8 +86,8 @@ pub struct Runtime {
 }
 
 impl Runtime {
-    pub async fn start(config: Config) -> Result<Runtime, Error> {
-        let (stop_tx, stop_rx) = oneshot::channel();
+    pub async fn start(config: Config) -> Result<(Runtime, CancellationToken), Error> {
```
[the remark this answers is not in the document]
Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:89]: One problem with `stop_wait` is that the runtime has to be moved to the tasks that handles the `SIGINT` and `SIGTERM`. In general, I think most of the time we just want to signal the shutdown without waiting for it. Separating the token from the runtime allows that. An other alternative that would be to send a `Request::Shutdown` instead of using the token.

File: northstar/src/runtime/mod.rs (hunk 4 above)
Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:134]: hmm I don't understand why `wait` has to be combined with `stop`.  Sometimes we want to stop without wait and sometimes we want to wait without stop.

File: northstar/src/runtime/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (flxo) [reply northstar/src/runtime/mod.rs:89]: I think there's no need for a separate task at all as it's done in the current implementation. That a `Request::Shutdown` is sent internally is a impl detail and should not be visible on the pub API.

File: northstar/src/runtime/mod.rs (hunk 4 above)
Comment (flxo) [reply northstar/src/runtime/mod.rs:134]: What's the use case for stopping without waiting for a proper shutdown?

Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:134]: For the signal handling for example. They just trigger the stop without waiting.

Comment (flxo) [reply northstar/src/runtime/mod.rs:134]: Today there's a wait ;-). I just don't get the point why this is bad.

Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:134]: hmmm I didn't understand that :confused: I don't understand why this is bad either

File: northstar/src/runtime/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:89]: Well separating these two things avoids to implement `Future` for `Runtime` and allows other type of "events" to trigger a shutdown if we so require in the future. I think this makes the handling of the runtime simpler and cleaner. Why is it better the way it was before?.

File: northstar/src/runtime/mod.rs (hunk 4 above)
Comment (alfonsoros88) [reply northstar/src/runtime/mod.rs:134]: Ah got it, yeah the wait is done by someone else. The one owning the runtime is the one waiting for it. That's related to the other discussion of course.
